src/coreclr/jit/morph.cpp

```diff
@@ -9309,15 +9309,8 @@ GenTree* Compiler::fgMorphCall(GenTreeCall* call)
         // BBJ_THROW would result in the tail call being dropped as the epilog is generated
         // only for BBJ_RETURN blocks.
         //
-        // Currently this doesn't work for non-void callees. Some of the code that handles
-        // fgRemoveRestOfBlock expects the tree to have GTF_EXCEPT flag set but call nodes
-        // do not have this flag by default. We could add the flag here but the proper solution
-        // would be to replace the return expression with a local var node during inlining
-        // so the rest of the call tree stays in a separate statement. That statement can then
-        // be removed by fgRemoveRestOfBlock without needing to add GTF_EXCEPT anywhere.
-        //
 
-        if (!call->IsTailCall() && call->TypeGet() == TYP_VOID)
+        if (!call->IsTailCall() && (call->TypeIs(TYP_VOID) || (call->gtFlags & GTF_EXCEPT)))
```

Comment:
I'm not sure why you added the extra check here -- did you see cases where it mattered?

If so, can you add back a comment explaining why?

Reply:
Ah, right it's not needed, but there were two asserts for thrownode to be always of VOID type - I've removed them and wrote a few tests to cover them.